From 7886f1c7cdea3758e5a60a004c0dd8fc261e8205 Mon Sep 17 00:00:00 2001
From: Simon Glass <sjg@chromium.org>
Date: Wed, 14 Jun 2017 21:28:46 -0600
Subject: [PATCH 0408/1200] dm: ahci: Add a driver for SCSI on AHCI

Some AHCI drivers use SCSI under the hood. Rather than making the AHCI
driver be in the SCSI uclass it makes sense to have the AHCI device create
a SCSI device as a child. That way we can handle any AHCI-specific
operations rather than trying to pretend tha the device is just SCSI.

To handle this we need to provide a way for AHCI drivers to bind a SCSI
device as its child, and probe it. Add functions for this.

Signed-off-by: Simon Glass <sjg@chromium.org>
Reviewed-by: Bin Meng <bmeng.cn@gmail.com>
---
 include/ahci.h | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/include/ahci.h b/include/ahci.h
index b3882d54cc..e5d95d8632 100644
--- a/include/ahci.h
+++ b/include/ahci.h
@@ -295,4 +295,26 @@ int ahci_probe_scsi_pci(struct udevice *ahci_dev);
  */
 int ahci_init_dm(struct udevice *dev, void __iomem *base);
 
+/**
+ * ahci_bind_scsi() - bind a new SCSI bus as a child
+ *
+ * Note that the SCSI bus device will itself bind block devices
+ *
+ * @ahci_dev: AHCI parent device
+ * @devp: Returns new SCSI bus device
+ * @return 0 if OK, -ve on error
+ */
+int ahci_bind_scsi(struct udevice *ahci_dev, struct udevice **devp);
+
+/**
+ * ahci_probe_scsi() - probe and scan the attached SCSI bus
+ *
+ * Note that the SCSI device will itself bind block devices for any storage
+ * devices it finds.
+ *
+ * @ahci_dev: AHCI parent device
+ * @return 0 if OK, -ve on error
+ */
+int ahci_probe_scsi(struct udevice *ahci_dev);
+
 #endif
-- 
2.22.0

